docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,318 @@
+   other Kubernetes objects in the same namespace as the HPA (and the
+   namespace itself)
+
+4. The HPA should make scaling on multiple metrics in a single HPA
+   possible and explicit (splitting metrics across multiple HPAs leads to
+   the possibility of fighting between HPAs)
+
+### Specification ###
+
+```go
+type HorizontalPodAutoscalerSpec struct {
+    // the target scalable object to autoscale
+    ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef"`
+
+    // the minimum number of replicas to which the autoscaler may scale
+    // +optional
+    MinReplicas *int32 `json:"minReplicas,omitempty"`
+    // the maximum number of replicas to which the autoscaler may scale
+    MaxReplicas int32 `json:"maxReplicas"`
+
+    // the metrics to use to calculate the desired replica count (the
+    // maximum replica count across all metrics will be used).  The
+    // desired replica count is calculated multiplying the ratio between
+    // the target value and the current value by the current number of
+    // pods.  Ergo, metrics used must decrease as the pod count is
+    // increased, and vice-versa.  See the individual metric source
+    // types for more information about how each type of metric
+    // must respond.
+    // +optional
+    Metrics []MetricSpec `json:"metrics,omitempty"`
+}
+
+// a type of metric source
+type MetricSourceType string
+var (
+    // a metric describing a kubernetes object (for example, hits-per-second on an Ingress object)
+    ObjectSourceType MetricSourceType = "object"
+    // a metric describing each pod in the current scale target (for example, transactions-processed-per-second).
+    // The values will be averaged together before being compared to the target value
+    PodsSourceType MetricSourceType = "pods"
+    // a resource metric known to Kubernetes, as specified in requests and limits, describing each pod
+    // in the current scale target (e.g. CPU or memory).  Such metrics are built in to Kubernetes,
+    // and have special scaling options on top of those available to normal per-pod metrics (the "pods" source)
+    ResourceSourceType MetricSourceType = "resource"
+)
+
+// a specification for how to scale based on a single metric
+// (only `type` and one other matching field should be set at once)
+type MetricSpec struct {
+    // the type of metric source (should match one of the fields below)
+    Type MetricSourceType `json:"type"`
+
+    // a metric describing a single kubernetes object (for example, hits-per-second on an Ingress object)
+    Object *ObjectMetricSource `json:"object,omitempty"`
+    // a metric describing each pod in the current scale target (for example, transactions-processed-per-second).
+    // The values will be averaged together before being compared to the target value
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
+    // a resource metric known to Kubernetes, as specified in requests and limits, describing each pod
```

Comment:
There are some resources that could be exported via resource metric API which are not CPU and memory and does not have requests & limits. Currently it might be `ResourceStorage` and `ResourceNvidiaGPU` (we already have resource names for those). I imagine there might be more that we expose via metric API but they would not be used in requests and limits.

I'd suggest to just drop this part of the sentence:

```
// a resource metric known to Kubernetes describing each pod in the current scale target (e.g. CPU or memory).
```